### PR TITLE
disk_format: New case qemu_img_convert_image_with_unaligned_size

### PR DESCRIPTION
**Pull request by hellohellenmao**
Add new case for 'qemu-img convert' image with unaligned size

ID: 1760292
Signed-off-by: Tingting Mao <timao@redhat.com>

### CODE REVIEW

File: qemu/tests/qemu_img_convert_image_with_unaligned_size.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+from virttest import data_dir
+from virttest.qemu_io import QemuIOSystem
+from virttest.qemu_storage import QemuImg
+
+
+def run(test, params, env):
+    """
+    qemu-img supports convert images with unaligned size.
+
+    1. create source image via truncate, and  write 10k "1" into
```
Comment (lolyu) [inline qemu/tests/qemu_img_convert_image_with_unaligned_size.py]: `preallocation=falloc` doesn't equal to `truncate`.

File: qemu/tests/qemu_img_convert_image_with_unaligned_size.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    2. convert the source image to target, check whether there is error
+
+    :param test: Qemu test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+    def _qemu_io(img, cmd):
+        """Run qemu-io cmd to a given img."""
+        logging.info("Run qemu-io %s" % img.image_filename)
+        QemuIOSystem(test, params, img.image_filename).cmd_output(cmd, 120)
```
Comment (lolyu) [inline qemu/tests/qemu_img_convert_image_with_unaligned_size.py]: You should check if there is an error.

File: qemu/tests/qemu_img_convert_image_with_unaligned_size.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    cache_mode = params["cache_mode"]
+    img_dir = data_dir.get_data_dir()
+
+    source = QemuImg(params.object_params(src_image), img_dir, src_image)
+    source.create(source.params)
+    fd = os.open(source.image_filename, os.O_RDWR | os.O_CREAT)
+    os.ftruncate(fd, int(params["truncate_size"]))
+    _qemu_io(source, 'write -P 1 0 %s' % params["write_size"])
+
+    source.convert(source.params, img_dir, cache_mode="none",
```
Comment (lolyu) [inline qemu/tests/qemu_img_convert_image_with_unaligned_size.py]: should `fail` the test if conversion errors.

**Comment by hellohellenmao** [comment]
(1/1) Host_RHEL.m8.u1.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.qemu_img_convert_image_with_unaligned_size: PASS (2.73 s)

**Comment by hellohellenmao** [comment]
(1/1) Host_RHEL.m8.u1.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.qemu_img_convert_image_with_unaligned_size: PASS (2.75 s)

**Comment by hellohellenmao** [comment]
@lolyu @zhenyzha Help to review, please.

File: qemu/tests/qemu_img_convert_image_with_unaligned_size.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+            QemuIOSystem(test, params, img.image_filename).cmd_output(cmd, 120)
+        except process.CmdError:
+            test.fail("qemu-io to '%s' failed." % img.image_filename)
+
+    src_image = params["images"]
+    tgt_image = params["convert_target"]
+    img_dir = data_dir.get_data_dir()
+
+    source = QemuImg(params.object_params(src_image), img_dir, src_image)
+    fd = open(source.image_filename, mode='wt')
```
Comment (lolyu) [inline qemu/tests/qemu_img_convert_image_with_unaligned_size.py]: better be:
```python
with open(source.image_filename, mode="wb"):
    pass
```

Comment (lolyu) [reply qemu/tests/qemu_img_convert_image_with_unaligned_size.py]: Should be binary mode.

**Comment by hellohellenmao** [comment]
@lolyu Review again, please~

File: qemu/tests/qemu_img_convert_image_with_unaligned_size.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            test.fail("qemu-io to '%s' failed." % img.image_filename)
+
+    src_image = params["images"]
+    tgt_image = params["convert_target"]
+    img_dir = data_dir.get_data_dir()
+
+    source = QemuImg(params.object_params(src_image), img_dir, src_image)
+    with open(source.image_filename, mode='wb') as fd:
+        fd.truncate(int(params["truncate_size"]))
+        _qemu_io(source, 'write -P 1 0 %s' % params["write_size"])
```
Comment (lolyu) [inline qemu/tests/qemu_img_convert_image_with_unaligned_size.py]: LINE 38 should not be included in the `with` block.

Comment (hellohellenmao) [reply qemu/tests/qemu_img_convert_image_with_unaligned_size.py]: Yes, thanks.

**Comment by hellohellenmao** [comment]
(1/1) Host_RHEL.m8.u1.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.qemu_img_convert_image_with_unaligned_size: PASS (2.83 s)

File: qemu/tests/cfg/qemu_img_convert_image_with_unaligned_size.cfg [hunk 6]
```diff
@@ -0,0 +1,17 @@
+- qemu_img_convert_image_with_unaligned_size:
+    only raw
+    virt_test_type = qemu
+    type = qemu_img_convert_image_with_unaligned_size
+    start_vm = no
+    required_qemu = [2.12.0-83,)
```
Comment (zhenyzha) [inline qemu/tests/cfg/qemu_img_convert_image_with_unaligned_size.cfg]: What about qemu-kvm-rhev, not in the scope of testing? like 2.12.0-38

Comment (hellohellenmao) [reply qemu/tests/cfg/qemu_img_convert_image_with_unaligned_size.cfg]: Yes, not in the scope. The related bug[1] is closed by our PM.

[1]https://bugzilla.redhat.com/show_bug.cgi?id=1649788

**Comment by zhenyzha** [comment]
(1/1) Host_RHEL.m8.u1.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.qemu_img_convert_image_with_unaligned_size: PASS (2.48 s)

LGTM.Acked-by: Zhenyu Zhang <zhenyzha@redhat.com>

**Pull request merged**